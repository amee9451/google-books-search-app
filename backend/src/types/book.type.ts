export interface ImageLinks {
  smallThumbnail?: string;
}

export interface VolumeInfo {
  title: string;
  authors?: string[];
  publishedDate?: string;
  description?: string;
  publisher?: string;
  imageLinks?: ImageLinks;
  previewLink?: string;
}

export interface SaleInfo {
  listPrice?: {
    amount?: number | string;
  };
}

export interface BookItem {
  volumeInfo: VolumeInfo;
  saleInfo: SaleInfo;
}
export interface BookApiResponse {
  id: string;
  volumeInfo: {
    title: string;
    authors?: string[];
    description?: string;
    imageLinks?: {
      thumbnail?: string;
      smallThumbnail?: string;
    };
    publishedDate?: string;
    [key: string]: any;
  };
  saleInfo: SaleInfo;
  totalItems: number;
}
export interface Stats {
  mostFrequentAuthor?: string;
  earliestPublishedYear?: string;
  latestPublishedYear?: string;
  totalItems?: number;
  responseTimeMs?: number;
}
export interface BooksApiResult {
  items: BookApiResponse[];
  totalItems: number;
}