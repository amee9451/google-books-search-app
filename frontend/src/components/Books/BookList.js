import { jsx as _jsx } from "react/jsx-runtime";
import BookItem from './BookItem';
const BookList = ({ books }) => {
    return (_jsx("div", { className: "space-y-4", children: books.map((book, index) => (_jsx(BookItem, { book: book }, index))) }));
};
export default BookList;
