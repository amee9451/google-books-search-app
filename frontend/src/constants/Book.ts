
export const UNNOWN_AUTHOR = 'Unknown Author';
export const FALLBACK_IMAGE_URL = 'https://placehold.co/100x100'
export const PRIVIEW_THIS_BOOK = 'Preview this book'
export const DESCRIPTION = "Description"
export const PAGINATION_LIST = [5,10, 20, 30, 40];
export const TOTAL_ITEMS='Total Results'
export const MOST_FREQUENT_AUTHOR='Most Frequent Author'
export const EARLIEST_PUBLISHED_YEAR='Earliest Publication'
export const LATEST_PUBLISHED_YEAR='Latest Publication'
export const RESPONSE_TIME_MS = 'Response Time'
export const BOOK_SEARCH_API = '/api/books/'
export const DEFAULT_PAGE_LIMIT=5
export const DEFAULT_PAGE_NUMBER = 1