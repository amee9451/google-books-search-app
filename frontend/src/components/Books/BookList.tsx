import React from 'react';
import BookItem from './BookItem';
import { BookListProps } from '../../types/Books/Book.types';


const BookList: React.FC<BookListProps> = ({ books }) => {
  return (    
    <div className="space-y-4">
      {books.map((book, index) => (
          <BookItem key={index} book={book} />
      ))}
    </div>
  );
};

export default BookList;
