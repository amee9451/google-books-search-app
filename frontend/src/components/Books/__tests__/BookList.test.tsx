import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import BookList from '../BookList';
import { Book } from '../../../types/Books/Book.types';
import SampleData from '../__mocks__/sampleBooks';

// Mock BookItem with proper typing
jest.mock('../BookItem', () => ({ 
  __esModule: true,
  default: ({ book }: { book: typeof SampleData[0] }) => (
    <div data-testid="book-item">
      {book.volumeInfo.title} {/* Removed authors */}
    </div>
  )
}));

describe('BookList Component', () => {
  it('renders book titles correctly', () => {
  render(<BookList books={SampleData} />);
  
  // Check for partial text matches
  expect(screen.getByText(/The Pragmatic Programmer/)).toBeInTheDocument();
  expect(screen.getByText(/Clean Code/)).toBeInTheDocument();
  
  // More precise checks using test-id and text content
  const bookItems = screen.getAllByTestId('book-item');
  expect(bookItems[0]).toHaveTextContent('The Pragmatic Programmer');
  expect(bookItems[1]).toHaveTextContent('Clean Code');
  });
it('renders book titles correctly once', () => {
  render(<BookList books={SampleData} />);
  SampleData.forEach(book => {
    const elements = screen.getAllByText(book.volumeInfo.title);
    expect(elements.length).toBe(1); 
  });
});
  
  it('renders book titles correctly', async () => {
  render(<BookList books={SampleData} />);
  
  await expect(screen.findByText('The Pragmatic Programmer')).resolves.toBeInTheDocument();
  await expect(screen.findByText('Clean Code')).resolves.toBeInTheDocument();
  });
  it('passes correct book data to BookItems', () => {
  render(<BookList books={SampleData} />);
  const bookItems = screen.getAllByTestId('book-item');
  
  expect(bookItems[0]).toHaveTextContent(SampleData[0].volumeInfo.title);
  expect(bookItems[1]).toHaveTextContent(SampleData[1].volumeInfo.title);
  });
  it('matches snapshot with books', () => {
    const { asFragment } = render(<BookList books={mockBooks} />);
    expect(asFragment()).toMatchSnapshot();
  });
  const mockBooks: Book[] = SampleData;
  it('displays correct number of books', () => {
    render(<BookList books={mockBooks} />);
    const bookItems = screen.getAllByTestId('book-item');
    expect(bookItems).toHaveLength(2);
  });
});