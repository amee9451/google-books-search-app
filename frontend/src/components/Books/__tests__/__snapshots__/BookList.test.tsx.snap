// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookList Component matches snapshot with books 1`] = `
<DocumentFragment>
  <div
    class="space-y-4"
  >
    <div
      data-testid="book-item"
    >
      The Pragmatic Programmer 
    </div>
    <div
      data-testid="book-item"
    >
      Clean Code 
    </div>
  </div>
</DocumentFragment>
`;
